all-sdk/autotuning.md

```diff
@@ -0,0 +1,431 @@
+  used_slots: Vec<WorkflowSlotInfo>,
+  /// Current size of the workflow cache.
+  num_cached_workflows: usize,
+  /// The limit on the size of the cache, if any. This is important for users to know as discussed below in the section
+  /// on workflow cache management.
+  max_cache_size: Option<usize>,
+  // ... Possibly also metric information
+}
+struct ActivitySlotsInfo {
+  used_slots: Vec<ActivitySlotInfo>,
+}
+struct LocalActivitySlotsInfo {
+  used_slots: Vec<LocalActivitySlotInfo>,
+}
+
+struct WorkflowSlotInfo {
+  workflow_type: String,
+  // task queue, worker id, etc...
+}
+struct ActivitySlotInfo {
+  activity_type: String,
+  // etc...
+}
+struct LocalActivitySlotInfo {
+  activity_type: String,
+  // etc...
+}
+
+struct WorkflowSlotKind {}
+struct ActivitySlotKind {}
+struct LocalActivitySlotKind {}
+trait SlotKind {
+  type Info;
+}
+impl SlotKind for WorkflowSlotKind {
+  type Info = WorkflowSlotInfo;
+}
+impl SlotKind for ActivitySlotKind {
+  type Info = ActivitySlotInfo;
+}
+impl SlotKind for LocalActivitySlotKind {
+  type Info = LocalActivitySlotInfo;
+}
+trait WorkflowTaskSlotSupplier: SlotSupplier<SlotKind=WorkflowSlotKind> {}
+trait ActivityTaskSlotSupplier: SlotSupplier<SlotKind=ActivitySlotKind> {}
+trait LocalActivityTaskSlotSupplier: SlotSupplier<SlotKind=LocalActivitySlotKind> {}
+
+/// Users might want to be able to pause the handing-out of slots as an effective way of pausing their workers.
+/// We can provide an implementation for this that wraps their implementation, or one of the defaults we provide.
+struct PauseableSlotSupplier<T: SlotKind> {
+  inner: Arc<dyn SlotSupplier<SlotKind=T>>,
+  paused: AtomicBool,
+  // ... details
+}
+impl<T: SlotKind> PauseableSlotSupplier<T> {
+  pub fn new(supplier: impl SlotSupplier<SlotKind=T>) -> Self { /* ... */ }
+  pub fn pause(&self) { /* ... */ }
+  pub fn resume(&self) { /* ... */ }
+}
+impl<T: SlotKind> SlotSupplier for PauseableSlotSupplier<T> {
+  // ... implementation which checks the paused flag before handing out a slot
+}
+```
+
+## Workflow cache management
+
+The workflow cache is maintained independently of the number of slots in use. There could be 0 slots in use and 100
+workflows cached, and that's a totally normal scenario. We might also choose to make this user-configurable, though we
+certainly don't need to have that right out of the gate.
+
+Keeping our fixed-size implementation to start is fine, because we won't ask the user's slot provider for a slot unless
+there is room in the cache (either additional capacity, or an idle cached workflow that can be evicted). This avoids
+a potential problem where we ask for a slot, they provide one, but there's nothing we can evict at the moment.
+
+If and when we do introduce user-configurable caching, it's likely that they would want the same thing to be responsible
+for both slots and cache management, since the two are closely related.
+
+A possible future:
+
+```rust
+trait WorkflowCacheSizer {
```

Comment:
I would like us to explore the idea of fully pluggable workflow cache instead of just a pluggable cache sizer. I suspect there is some performance wins for certain workflows to use different caching strategy then we currently use. I see you cover some of it with the `eviction_hint `, but I think just letting users provide a cache implementations would be simpler .

Reply:
sorry if I'm late to the game here, but what do you mean by "fully pluggable"?  Do you (basically) mean user-written, or that we should provide multiple implementations?  If so could you give a hint what a different (from current) version might look like?

Reply:
By fully pluggable I mean user written and we provide a few implementation. Right now we have a bounded LRU cache, a different version I would like to experiment with is a cache that accounts for how much work it would take to replay the workflow. One way I image it could do that is when it needs to evict a workflow weight the history size in what workflow we evict.

Reply:
Concur, the difference between providing a cache interface and a can-cache interface is minimal, but the former has lots of upside. Let's just expose a get/put/delete cache interface that also has a "can allow" call.

For core langs however, they will only expose the "can" (and eviction callbacks if they want) because the stateful cache impl has to be in Rust IMO w/ just hooks for checking.

Reply:
I don't want a provided implementation to be literally responsible for storing the data, though, that seems potentially easy to create confusing-to-diagnose bugs & more work for the implementor.

What would actual storage provide that `eviction_hint` doesn't? The example you gave @Quinn-With-Two-Ns seems to be serviced by the eviction hint.

Reply:
The only reason is flexibility. It's basically a question of allowing replacement of the cache impl vs just the cache size. The former is more flexible I think and easy enough to build a sizer on top of, but I guess you may have to abstract the sizer anyways then. So a double-abstraction could exist: a worker accepts a custom cache implementation, and the fixed-size implementation could accept a dynamic sizer.

I don't have a strong opinion and can understand why just the sizer is enough. But I would change the names on some of these things a bit.

Reply:
> but I think interceptors and data converters have equal if not more potential for introducing bugs.
This is true, but, also doesn't supply a reason to introduce _additional_ dangerous things. Those are necessary, but the user actually storing the cached data for us certainly isn't.